node/src/components/consensus/protocols/highway/tests/consensus_environment.rs

```diff
@@ -0,0 +1,434 @@
+    ) -> Self {
+        let mut highway = new_test_highway_protocol_with_era_height(
+            validators
+                .iter()
+                .map(|(pub_key, value)| (pub_key.clone(), value.1)),
+            vec![],
+            Some(10),
+        );
+        // our active validator will be the first in the map
+        let (pub_key, (keypair, _)) = validators.iter().next().unwrap();
+        // this is necessary for the round exponent to be tracked - it only happens in the
+        // ActiveValidator
+        let _ =
+            highway.activate_validator(pub_key.clone(), keypair.clone(), Timestamp::zero(), None);
+        Self {
+            highway,
+            leaders: LeaderSequence::new(
+                0, // used as the seed in `new_test_highway_protocol`
+                &validators
+                    .values()
+                    .map(|(_keypair, weight)| (*weight).into())
+                    .collect(),
+                vec![true; validators.len()].into(),
+            ),
+            validators,
+            current_round_start: 0,
+            slow_validators,
+            rng: NodeRng::new(),
+            finalized_blocks: vec![],
+        }
+    }
+
+    fn highway(&self) -> &HighwayProtocol<ClContext> {
+        self.highway.as_any().downcast_ref().unwrap()
+    }
+
+    fn our_pub_key(&self) -> &PublicKey {
+        self.validators.keys().next().unwrap()
+    }
+
+    fn is_slow(&self) -> bool {
+        self.slow_validators.contains(self.our_pub_key())
+    }
+
+    fn min_round_len(&self) -> u64 {
+        self.highway()
+            .highway()
+            .state()
+            .params()
+            .min_round_length()
+            .millis()
+    }
+
+    fn clone_state(&self) -> State<ClContext> {
+        self.highway().highway().state().clone()
+    }
+
+    pub(super) fn crank_round(&mut self) {
+        let min_round_len = self.min_round_len();
+        let round_id = Timestamp::from(self.current_round_start);
+        let leader = self.leaders.leader(round_id.millis());
+        let leader_pub_key = self
+            .validators
+            .keys()
+            .nth(leader.0 as usize)
+            .unwrap()
+            .clone();
+        let leader_is_slow = self.slow_validators.contains(&leader_pub_key);
+
+        let pre_proposal_state = self.clone_state();
+
+        let (mut post_proposal_state, maybe_proposal_msg) = if leader.0 == 0 {
+            // our active validator is the proposer
+            self.this_node_propose();
+            (self.clone_state(), None)
+        } else {
+            // another validator is the proposer
+            self.other_node_propose(leader)
+        };
+
+        // if we're slow, we're going to create a witness unit before receiving any units from
+        // other nodes, effectively not citing any of them
+        // if we're the leader, our proposal and confirmation are already in the state at this
+        // point
+        if self.is_slow() {
```

Comment:
I'm a bit stumped as to what this does that is supposed to model the 0th validator being slow. The logic is exactly the same as it is on lines 176-183, when the 0th validator is NOT slow?

Reply:
Yes, the difference is that here it is executed on the state that doesn't contain the units from other validators yet (because we simulate them not having yet arrived), and in lines 176-183 it is executed on the state that contains those units (it's after the units have been processed) - because if we aren't slow, we are able to receive those units on time.